### PR TITLE
Add functionality to update user roles and groups

### LINKED ISSUE #591: User management - Add functionality to update user role list
**Describe the solution you'd like**
The capability to remove already assigned user roles or assign new roles to the user will be introduced with this feature.

### PR DESCRIPTION
**Pull request by NipuniBhagya**
## Preview
image

image

### CODE REVIEW

File: apps/admin-portal/src/components/users/user-groups-edit.tsx [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+                }
+            });
+            setPrimaryGroupsList(groupsMap);
+        }
+    };
+
+    /**
+     * The following function remove already assigned roles from the initial roles.
+     *
+     * @param domain
```
Comment (inthirakumaaran) [inline apps/admin-portal/src/components/users/user-groups-edit.tsx:118]: add param descriptions

File: apps/admin-portal/src/components/users/user-groups-edit.tsx [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+    const [ tempGroupList, setTempGroupList ] = useState([]);
+    const [ initialGroupList, setInitialGroupList ] = useState([]);
+    const [ initialTempGroupList, setInitialTempGroupList ] = useState([]);
+    const [ primaryGroups, setPrimaryGroups ] = useState(undefined);
+    const [ primaryGroupsList, setPrimaryGroupsList ] = useState<Map<string, string>>(undefined);
+    const [ checkedUnassignedListItems, setCheckedUnassignedListItems ] = useState<RolesMemberInterface[]>([]);
+    const [ checkedAssignedListItems, setCheckedAssignedListItems ] = useState<RolesMemberInterface[]>([]);
+    const [ isSelectUnassignedRolesAllRolesChecked, setIsSelectUnassignedAllRolesChecked ] = useState(false);
+    const [ isSelectAssignedAllRolesChecked, setIsSelectAssignedAllRolesChecked ] = useState(false);
+    const [ assignedGroups, setAssignedGroups ] = useState([]);
```
Comment (DimalChandrasiri) [inline apps/admin-portal/src/components/users/user-groups-edit.tsx:71]: Better to add the types for the state objects.

File: apps/admin-portal/src/components/users/user-groups-edit.tsx [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+                .then(() => {
+                    onAlertFired({
+                        description: "Assigning new groups for the user successful",
+                        level: AlertLevels.SUCCESS,
+                        message: "Update user groups successful"
+                    });
+                    handleCloseAddNewGroupModal();
+                    handleUserUpdate(user.id);
+                })
+                .catch((error) => {
```
Comment (inthirakumaaran) [inline apps/admin-portal/src/components/users/user-groups-edit.tsx:381]: Shall we make use of the error here?

File: apps/admin-portal/src/components/users/user-groups-edit.tsx [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+            setCheckedAssignedListItems(checkedGroups);
+        } else {
+            checkedGroups.push(group);
+            setCheckedAssignedListItems(checkedGroups);
+        }
+    };
+
+    const addGroups = () => {
+        const addedRoles = [ ...tempGroupList ];
+        if (checkedUnassignedListItems?.length > 0) {
```
Comment (DimalChandrasiri) [inline apps/admin-portal/src/components/users/user-groups-edit.tsx:185]:
```suggestion

        if (checkedUnassignedListItems?.length > 0) {
```

File: apps/admin-portal/src/components/users/user-groups-edit.tsx [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+        setTempGroupList(addedRoles);
+        setInitialTempGroupList(addedRoles);
+        setGroupList(groupList.filter(x => !addedRoles?.includes(x)));
+        setInitialGroupList(groupList.filter(x => !addedRoles?.includes(x)));
+        setIsSelectUnassignedAllRolesChecked(false);
+    };
+
+    const removeGroups = () => {
+        const removedRoles = [ ...groupList ];
+        if (checkedAssignedListItems?.length > 0) {
```
Comment (DimalChandrasiri) [inline apps/admin-portal/src/components/users/user-groups-edit.tsx:201]:
```suggestion

        if (checkedAssignedListItems?.length > 0) {
```

File: apps/admin-portal/src/components/users/user-roles.tsx [hunk 6]
```diff
@@ -53,171 +73,121 @@ export const UserRolesList: FunctionComponent<UserRolesPropsInterface> = (
...
         }
-    };
+        setAssignedRoles(user.groups);
+        mapUserRoles();
+    }, []);
 
     /**
-     * The following function enables the user to deselect all the roles
-     * at once.
+     * The following useEffect will be triggered when the
```
Comment (inthirakumaaran) [inline apps/admin-portal/src/components/users/user-roles.tsx:124]: Can't we have them in one line

File: apps/admin-portal/src/components/users/wizard/add-user-wizard.tsx [hunk 7]
```diff
@@ -90,40 +90,18 @@ export const AddUserWizard: FunctionComponent<AddUserWizardPropsInterface> = (
...
 
     const [ groupList, setGroupsList ] = useState([]);
     const [ tempGroupList, setTempGroupList ] = useState([]);
     const [ initialGroupList, setInitialGroupList ] = useState([]);
+    const [ initialTempGroupList, setInitialTempGroupList ] = useState([]);
 
     const [ applicationRoles, setApplicationRoles ] = useState([]);
     const [ internalRoles, setInternalRoles ] = useState([]);
     const [ isInternalRolesSet, setIsInternalRolesSet ] = useState(false);
     const [ isApplicationRolesSet, setIsApplicationRolesSet ] = useState(false);
```
Comment (DimalChandrasiri) [inline apps/admin-portal/src/components/users/wizard/add-user-wizard.tsx:103]: Add the type for the state objects

**Pull request merged**